dicts/dicts.gobra

```diff
@@ -0,0 +1,605 @@
+decreases
+pure func NotDisjointSharedKey(d1, d2 dict[int]int) util.Unit {
+	return util.Unit{}
+}
+
+// There is only one empty map.
+ghost
+requires IsEmpty(d)
+ensures d == Empty()
+decreases
+pure func EmptyIsUnique(d dict[int]int) util.Unit {
+	return util.Unit{}
+}
+
+// An empty dictionary contains no keys.
+ghost
+requires IsEmpty(d)
+ensures !(x in domain(d))
+decreases
+pure func NotInEmpty(d dict[int]int, x int) util.Unit {
+	return util.Unit{}
+}
+
+// There exists a key in a non-empty dictionary.
+ghost
+requires !IsEmpty(d)
+ensures exists x int :: x in domain(d)
+decreases
+pure func NotEmptyKeyExists(d dict[int]int, x int) util.Unit {
+	return util.Unit{}
+}
+
+// If a key is in d, d is not empty.
+ghost
+requires x in domain(d)
+ensures !IsEmpty(d)
+decreases
+pure func KeyInDomainDictNotEmpty(d dict[int]int, x int) util.Unit {
+	return util.Unit{}
+}
+
+// Equality -> ExtEqual
+// First update axiom: UpdateSame, UpdateDifference
+
+// Inserting a new key increases the cardinality of the dictionary by 1.
+// Updating the value of a key does not change the cardinality.
+ghost
+ensures x in domain(d) ==> len(d[x = y]) == len(d)
+ensures !(x in domain(d)) ==> len(d[x = y]) == len(d) + 1
+decreases
+pure func InsertUpdateLen(d dict[int]int, x, y int) util.Unit {
+	return util.Unit{}
+}
+
+// If a value is in the range of a dictionary, there exists a corresponding key.
+ghost
+requires y in range(d)
+ensures exists x int :: x in domain(d) && d[x] == y
+decreases
+pure func ValueHasKey(d dict[int]int, y int) util.Unit {
+	return util.Unit{}
+}
+
+// If a value is in the domain of a dictionary, the corresponding value is in its range.
+ghost
+requires x in domain(d)
+ensures d[x] in range(d)
+decreases
+pure func KeyMapsToRange(d dict[int]int, x int) util.Unit {
+	return util.Unit{}
+}
+
+// Why3 (fmap)
+// ==, extensionality -> ExtEqual
+// mem k m -> k in domain(m)
+// mapsto k v m -> k in domain(m) && m[k] == v
+// mem_mapsto: skip (doesn't seem useful)
+// is_empty -> IsEmpty
+// mk, mk_*: skip (construction doesn't apply)
+// empty, is_empty_empty -> Empty
+// [<-], add k v m -> m[k = v]
+// add_contents_k -> UpdateSame
+// add_contents_other -> UpdateDifferent
+// add_domain -> InsertDomain
+// find k, find_def -> m[k]
+// remove k m -> Remove(m, k)
+// remove_contents -> RemoveDifferent
+// remove_domain -> RemoveDomain
+// size m -> len(m)
+
+// NOTE fmap (applicative maps) may be interesting when implementing gomaps
+
+// Why3 (map)
+// [], get f x -> f[x]
+// [<-], set f x v -> f[x = v]
+// const: skip (cannot come up with feasible implementation)
+// ==, extensionality -> ExtEqual
+// sorted_sub -> IsMonotonicFromTo
+
+// True iff a dictionary is monotonic. Only considers keys in the interval [start, end).
+ghost
+opaque
+decreases
+pure func IsMonotonicFromTo(d dict[int]int, start, end int) bool {
+	return forall x, y int :: {d[x], d[y]} (x in domain(d) && y in domain(d) && start <= x && x <= y && y < end) ==>
```

Comment:
This one is a bit awkward to break up, since `(x in domain(d) ...` must at least begin as the same line as the forall; otherwise the parser complains. Please let me know if the new version (upcoming commit) needs more work.